homeassistant/components/alexa/smart_home.py

```diff
@@ -1489,11 +1523,22 @@ def temperature_from_object(hass, temp_obj, interval=False):
     mode = request[API_PAYLOAD]['thermostatMode']
     mode = mode if isinstance(mode, str) else mode['value']
 
+    if mode not in API_THERMOSTAT_MODES:
+        msg = 'The requested thermostat mode {} is not supported'.format(mode)
+        return api_error(
+            request,
+            namespace='Alexa.ThermostatController',
+            error_type='UNSUPPORTED_THERMOSTAT_MODE',
+            error_message=msg
+        )
+
     operation_list = entity.attributes.get(climate.ATTR_OPERATION_LIST)
-    ha_mode = next(
-        (k for k, v in API_THERMOSTAT_MODES.items() if v == mode),
+    op_candidates = API_THERMOSTAT_MODES[mode]
+    ha_mode = mode.lower() if mode.lower() in op_candidates else next(
```

Comment:
The modes for alexa should be fixed, we shouldn't have to lowercase anything?

Reply:
The mode from Alexa is either of `HEAT`, `COOL`, `OFF` or `AUTO`, and all of them are upper case. 
I wanted to map those correctly to HA mode which is in lower case.

Reply:
That makes it very unreadable. The old method is better, we just need to make sure that `API_THERMOSTAT_MODES` is an ordered dict so that Python 3.5 will also map OFF correctly.

Reply:
So if we change the `API_THERMOSTAT_MODES` to an ordered dict, we can control which mapping is done first. That is the correct solution.